src/proc.cpp

```diff
@@ -568,22 +567,43 @@ static bool process_clean_after_marking(bool allow_interactive) {
             }
             proc_fire_event(L"JOB_EXIT", event_type_t::job_exit, j->job_id, 0);
 
-            itr = jobs().erase(itr);
-            erased = true;
+            erase_list.push_back(j.get());
         } else if (j->is_stopped() && !j->get_flag(job_flag_t::NOTIFIED)) {
             // Notify the user about newly stopped jobs.
             if (!j->get_flag(job_flag_t::SKIP_NOTIFICATION)) {
-                print_job_status(j, JOB_STOPPED);
+                print_job_status(j.get(), JOB_STOPPED);
                 found = true;
             }
             j->set_flag(job_flag_t::NOTIFIED, true);
         }
     }
 
-    if (found) fflush(stdout);
+    if (!erase_list.empty()) {
```

Comment:
This is fine but keep in mind that the number of jobs is typically small (almost always just one) and the number of jobs reaped in this function call is probably even smaller, so we don't need to be too clever here.